### PR TITLE
Add a lint for assert_throws and promise_rejects

### PR DESCRIPTION
**Pull request by jgraham**
This adds a rather large number of existing uses to the allow list

### CODE REVIEW

File: lint.whitelist [hunk 1]
```diff
@@ -757,3 +757,1407 @@ TESTHARNESS-IN-OTHER-TYPE: svg/extensibility/foreignObject/foreign-object-circul
 TESTHARNESS-IN-OTHER-TYPE: svg/extensibility/foreignObject/foreign-object-under-clip-path-crash.html
 TESTHARNESS-IN-OTHER-TYPE: svg/extensibility/foreignObject/foreign-object-under-defs-crash.html
 TESTHARNESS-IN-OTHER-TYPE: svg/svg-in-svg/svg-in-svg-circular-filter-reference-crash.html
+
+# Unconvverted tests using PROMISE_REJECTS
```
Comment (zcorpan) [inline lint.whitelist]:
```suggestion
# Unconverted tests using PROMISE_REJECTS
```

File: lint.whitelist [hunk 2]
```diff
@@ -757,3 +757,1407 @@ TESTHARNESS-IN-OTHER-TYPE: svg/extensibility/foreignObject/foreign-object-circul
...
+PROMISE_REJECTS:webxr/xrDevice_requestSession_immersive_unsupported.https.html
+PROMISE_REJECTS:webxr/xrDevice_requestSession_requiredFeatures_unknown.https.html
+PROMISE_REJECTS:webxr/xrSession_features_deviceSupport.https.html
+PROMISE_REJECTS:webxr/xrSession_prevent_multiple_exclusive.https.html
+PROMISE_REJECTS:webxr/xrSession_requestReferenceSpace.https.html
+PROMISE_REJECTS:webxr/xrSession_requestReferenceSpace_features.https.html
+PROMISE_REJECTS:webxr/xrSession_viewer_availability.https.html
+PROMISE_REJECTS:worklets/resources/import-tests.js
+
+# Unconvverted tests using ASSERT_THROWS
```
Comment (zcorpan) [inline lint.whitelist]:
```suggestion
# Unconverted tests using ASSERT_THROWS
```

File: docs/writing-tests/testharness-api.md [hunk 3]
```diff
@@ -213,10 +216,25 @@ function bar() {
...
+
+`type` - The expected exception name or code.  See the table of names
+and codes at https://heycam.github.io/webidl/#dfn-error-names-table If
+a number is passed it should be one of the numeric code values in that
+table (e.g. 3, 4, etc).  If a string is passed it can either be an
+exception name (e.g. "HierarchyRequestError", "WrongDocumentError") or
+the name of the corresponding error code
+(e.g. "HIERARCHY_REQUEST_ERR", "WRONG_DOCUMENT_ERR").
+
+`func` - func Function which should throw.
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
`promise` - a promise that should reject.
```

File: docs/writing-tests/testharness-api.md [hunk 4]
```diff
@@ -213,10 +216,25 @@ function bar() {
...
 promise_test(function(t) {
-  return promise_rejects(t, new TypeError(), bar());
+  return promise_rejects_js(t, new TypeError(), bar());
 }, "Another example");
 ```
 
+```js
+promise_rejects_dom(test_object, expected, promise, description)
+```
+assert the promise rejects with a DOMException of the expected type.
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
assert that the promise rejects with a DOMException of the expected type.
```

File: docs/writing-tests/testharness-api.md [hunk 5]
```diff
@@ -195,14 +195,17 @@ test may begin to execute before the returned promise has settled. Use
...
-promise_rejects(test_object, code, promise, description)
+promise_rejects_js(test_object, constructor, promise, description)
 ```
+assert that the promise rejects with a JS Error with the specified constructor.
 
-The `code` argument is equivalent to the same argument to the `assert_throws`
-function.
+`constructor` - the expected exception constructor
+
+`func` - a function that should throw
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
`promise` - a promise that should reject
```

(I think? I don't see a 'func' argument?)

File: docs/writing-tests/testharness-api.md [hunk 6]
```diff
@@ -213,10 +216,25 @@ function bar() {
...
+  return promise_rejects_js(t, new TypeError(), bar());
 }, "Another example");
 ```
 
+```js
+promise_rejects_dom(test_object, expected, promise, description)
+```
+assert the promise rejects with a DOMException of the expected type.
+
+`type` - The expected exception name or code.  See the table of names
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
`expected` - The expected exception name or code.  See the table of names
```

(Or rename 'expected' above?)

File: docs/writing-tests/testharness-api.md [hunk 7]
```diff
@@ -213,10 +216,25 @@ function bar() {
...
+and codes at https://heycam.github.io/webidl/#dfn-error-names-table If
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
and codes at https://heycam.github.io/webidl/#dfn-error-names-table. If
```

File: docs/writing-tests/testharness-api.md [hunk 8]
```diff
@@ -854,20 +872,26 @@ attribute attribute_name following the conditions specified by WebIDL
...
-             "TIMEOUT_ERR")
-  * object - asserts that the thrown exception must be any other kind
-             of exception, with a property called "name" that matches
-             `code.name`.
+`constructor` - the expected exception constructor
 
 `func` - a function that should throw
 
+### `assert_throws_dom(type, func, description)`
+assert a DOMException with the expected type is thrown.
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
assert that a DOMException with the expected type is thrown.
```

File: docs/writing-tests/testharness-api.md [hunk 9]
```diff
@@ -854,20 +872,26 @@ attribute attribute_name following the conditions specified by WebIDL
...
+
+`type` - The expected exception name or code.  See the table of names
+and codes at https://heycam.github.io/webidl/#dfn-error-names-table If
```
Comment (stephenmcgruer) [inline docs/writing-tests/testharness-api.md]:
```suggestion
and codes at https://heycam.github.io/webidl/#dfn-error-names-table. If
```

File: tools/lint/rules.py [hunk 10]
```diff
@@ -462,3 +462,19 @@ class TrailingWhitespaceRegexp(Regexp):
     description = "Whitespace at EOL"
     pattern = b"[ \t\f\v]$"
     to_fix = """Remove trailing whitespace from all lines in the file."""
+
+
+class AssertThrowsRegexp(Regexp):
+    pattern = br"assert_throws\("
+    name = "ASSERT_THROWS"
+    file_extensions = [".html", ".htm", ".js", ".xht", ".xhtml", ".svg"]
+    description = "Test-file line has a `assert_throws(...)` call"
```
Comment (stephenmcgruer) [inline tools/lint/rules.py]:
```suggestion
    description = "Test-file line has an `assert_throws(...)` call"
```

**Review by stephenmcgruer — APPROVED** [review]
LGTM with some nits (please address before landing)

**Comment by gsnedders** [comment]
> This adds a rather large number of existing uses to the allow list

What is the performance impact of this? Historically adding large numbers of items here has been problematic, given we scan through the entire list each time.

**Comment by jgraham** [comment]
> What is the performance impact of this?

I'm not sure although since @bzbarsky is actively working through the backlog, I'm also not that worried. I agree that the code here highly suboptimal…

**Comment by bzbarsky** [comment]
Basically, once https://github.com/web-platform-tests/wpt/pull/21286 merges I'll be able to put up my pr to get rid of most of the callsites for `assert_throws`.  Once that merges I'll poke at `promise_rejects`.

File: docs/writing-tests/testharness-api.md [hunk 11]
```diff
@@ -213,10 +216,25 @@ function bar() {
 }
 
 promise_test(function(t) {
-  return promise_rejects(t, new TypeError(), bar());
+  return promise_rejects_js(t, new TypeError(), bar());
```
Comment (domenic) [inline docs/writing-tests/testharness-api.md]: Shouldn't this be 
```suggestion
  return promise_rejects_js(t, TypeError, bar());
```

now?

**Comment by stephenmcgruer** [comment]
@jgraham With Boris' many changes over the last few days, I think the files added to the lint can be pared down significantly here now! :)

**Comment by jgraham** [comment]
So FWIW I measured the perf impact of this and we're so comically slow parsing HTML that nothing else makes a meaningful difference. With an updated list with bz's changes for `assert_throws` we spend < 2% of the runtime checking for allowed failures

**Review by stephenmcgruer — APPROVED** [review]
Lgtm, though Boris may have changed your whitelist yet again with their latest promise_rejects PR ;)

**Pull request merged**